lib/md051.js

```diff
@@ -0,0 +1,50 @@
+// @ts-check
+
+"use strict";
+
+const { filterTokens, addError, forEachHeading } = require("../helpers");
+
+/**
+ * Slugifies strings.
+ *
+ * @param {string} string The string to slugify.
+ * @returns {string} The slugified string.
+ */
+function slugify(string) {
+  return string
+    .toLowerCase()
+    .replace(/[.,/#!$%^&*;:{}[\]=_`~()]/g, "")
+    .split(" ")
+    .join("-");
+}
+
+module.exports = {
+  "names": [ "MD051", "no-dead-hash-links-within-document" ],
+  "description": "No dead hash links within document",
+  "tags": [ "links" ],
+  "function": function MD051(params, onError) {
+    const headings = [];
+    forEachHeading(params, (_heading, content) => {
+      headings.push("#" + slugify(content));
+    });
+    filterTokens(params, "inline", (token) => {
+      token.children.forEach((child) => {
+        const { lineNumber, type, attrs } = child;
+        if (type === "link_open") {
+          const href = attrs.find((attr) => attr[0] === "href");
+          if (href !== undefined &&
+              typeof href[1] === "string" &&
```

Comment:
Is this necessary? What else would it be?

Reply:
Indeed it is not necessary. My mistake, I would rather work with TypeScript ah ah.